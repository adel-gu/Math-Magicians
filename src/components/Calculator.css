button {
  border: none;
  padding: 15px 10px;
  background-color: aliceblue;
}

.calculator {
  width: 450px;
  max-width: 90%;
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  gap: 1px;
  box-shadow: 10px 8px 28px -7px rgba(0, 0, 0, 0.15);
}

.screen {
  background: var(--clr-screen);

  grid-column: span 4;
  padding: 5px;
  text-align: right;
}

.numberZero {
  grid-column: span 2;
}

.orang {
  background-color: var(--clr-orange);
  color: white;
}

button:hover {
  cursor: pointer;
  background: rgba(0, 0, 0, 0.1);
}

button.orang:hover {
  background-color: rgba(245, 145, 62, 0.8);
}
