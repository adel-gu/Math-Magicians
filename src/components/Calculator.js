import '../App.css';

const Screen = (props) => {
  return <span className='screen'>{props.result}</span>;
};

const Oprand = ({ data, className }) => {
  return <button className={className}>{data}</button>;
};

const Number = ({ value, className }) => {
  return <button className={className}>{value}</button>;
};

// Calculator

const Calculator = () => {
  return (
    <div className='calculator'>
      <Screen result='0' />
      {/* First row */}
      <Oprand data='AC' />
      <Oprand data='+/-' />
      <Oprand data='%' />
      <Oprand data='+' className='orang' />
      {/* Seccond row */}
      <Number value='7' />
      <Number value='8' />
      <Number value='9' />
      <Oprand data='*' className='orang' />
      {/* 3rd row */}
      <Number value='4' />
      <Number value='5' />
      <Number value='6' />
      <Oprand data='-' className='orang' />
      {/* 4th row */}
      <Number value='1' />
      <Number value='2' />
      <Number value='3' />
      <Oprand data='+' className='orang' />
      {/* 5th row */}
      <Number value='0' className='numberZero' />
      <Number value='.' />
      <Oprand data='=' className='orang' />
    </div>
  );
};

export default Calculator;
