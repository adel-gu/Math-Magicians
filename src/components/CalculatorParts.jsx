import PropTypes from 'prop-types';
// Sub Components => {Screen, oprands, Numbers}
const Screen = ({ result }) => <span className="screen">{result}</span>;
Screen.propTypes = { result: PropTypes.string.isRequired };

// Oprands buttons components
const Oprand = ({ data, className, onclick}) => (
  <button type="button" className={className} onClick={onclick}>
    {data}
  </button>
);
Oprand.propTypes = { data: PropTypes.string.isRequired };
Oprand.propTypes = { className: PropTypes.string };
Oprand.propTypes = { onclick: PropTypes.func.isRequired };

// Numbers buttons components
const Number = ({ value, className , onclick}) => (
  <button type="button" className={className} onClick={onclick}>
    {value}
  </button>
);
Number.propTypes = { value: PropTypes.string.isRequired };
Number.propTypes = { className: PropTypes.string};
Number.propTypes = { onclick: PropTypes.func.isRequired};

export {Screen, Oprand, Number};