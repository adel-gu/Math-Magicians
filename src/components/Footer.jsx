const Footer = () => {
  const mb = {
    marginBottom: '10px',
  };
  return (
    <footer className="footer">
      <div className="container header-footer py-2" style={mb}>
        <span>
          Build by
          <a href="https://github.com/adel-gu">Adel Guitoun ❤️</a>
        </span>
      </div>
    </footer>
  );
};

export default Footer;
