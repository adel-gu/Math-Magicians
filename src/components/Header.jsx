import { Link } from "react-router-dom";
// StyleSheets
import "../styles/Header.css";

const Header = () => {
  const FooterBackground = {
    backgroundColor: "red",
    marginTop: '10px',
    borderRadius: "100px",
    paddingInline: "1em"
  }
  return (
    <header className="header">
      <div className="container flex justify-between align-center" style={FooterBackground}>
        <a href="/" className="logo"><h3>Math Magicians</h3></a>
        <nav>
          <ul className="flex gap-3">
            <li><Link to="/" className="nav-link">Home</Link></li>
            <li><Link to="/calculator" className="nav-link">Calculator</Link></li>
            <li><Link to="/quote" className="nav-link">Quote</Link></li>
          </ul>
        </nav>
      </div>
    </header>
  )
}

export default Header;