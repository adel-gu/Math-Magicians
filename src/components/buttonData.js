const buttonsData = [
  { data: 'AC', columnPos: 1, rowPos: 2 },
  { data: '+/-', columnPos: 2, rowPos: 2 },
  { data: '%', columnPos: 3, rowPos: 2 },
  {
    data: '+',
    columnPos: 4,
    rowPos: 2,
    color: 'orang',
  },
  { data: '7', columnPos: 1, rowPos: 3 },
  { data: '8', columnPos: 2, rowPos: 3 },
  { data: '9', columnPos: 3, rowPos: 3 },
  {
    data: 'x',
    columnPos: 4,
    rowPos: 3,
    color: 'orang',
  },
  { data: '4', columnPos: 1, rowPos: 4 },
  { data: '5', columnPos: 2, rowPos: 4 },
  { data: '6', columnPos: 3, rowPos: 4 },
  {
    data: '-',
    columnPos: 4,
    rowPos: 4,
    color: 'orang',
  },
  { data: '1', columnPos: 1, rowPos: 5 },
  { data: '2', columnPos: 2, rowPos: 5 },
  { data: '3', columnPos: 3, rowPos: 5 },
  {
    data: '/',
    columnPos: 4,
    rowPos: 5,
    color: 'orang',
  },
  {
    data: '0',
    columnPos: 1,
    rowPos: 6,
    width: 2,
  },
  { data: '.', columnPos: 3, rowPos: 6 },
  {
    data: '=',
    columnPos: 4,
    rowPos: 6,
    color: 'orang',
  },
];

export default buttonsData;
