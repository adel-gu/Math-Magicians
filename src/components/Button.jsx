import React from 'react';
import PropTypes from 'prop-types';

// Stylesheets
import '../styles/Calculator.css';

export const Button = ({
  data,
  width,
  color,
  columnPos,
  rowPos,
  getBtnData,
}) => {
  const btnStyle = {
    gridColumn: `${columnPos} / span ${width}`,
    gridRow: rowPos,
  };
  return (
    <button
      style={btnStyle}
      className={`btn ${color}`}
      onClick={() => getBtnData(data)}
    >
      {data}
    </button>
  );
};

Button.defaultProps = {
  width: 1,
};

Button.propTypes = {
  data: PropTypes.string.isRequired,
  width: PropTypes.number,
  color: PropTypes.string,
  columnPos: PropTypes.number,
  rowPos: PropTypes.number,
  getBtnData: PropTypes.func.isRequired,
};
