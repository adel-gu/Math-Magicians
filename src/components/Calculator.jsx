import React, { useState } from 'react';
import { Button } from './Button';
import { Screen } from './Screen';
import buttonsData from './buttonData';
import calculate from '../logic/calculate';

// Stylesheets
import '../styles/Calculator.css';

export const Calculator = () => {
  let obj = {
    total: null,
    next: null,
    operation: null,
  };

  const [calObj, setCalcObj] = useState(obj);
  const [result, setResult] = useState('0');

  const getBtnData = (data) => {
    obj = calculate(calObj, data);
    setCalcObj(obj);
    let screenResult =
      `${obj.total ?? ''}${obj.operation ?? ''}${obj.next ?? ''}` || '0';
    setResult(screenResult);
  };

  return (
    <div className="calculator">
      <Screen result={result} />
      {buttonsData.map((btnData) => (
        <Button
          key={`${btnData.data}`}
          data={btnData.data}
          width={btnData.width}
          color={btnData.color}
          columnPos={btnData.columnPos}
          rowPos={btnData.rowPos}
          isOperation={btnData.isOperation}
          getBtnData={getBtnData}
        />
      ))}
    </div>
  );
};
