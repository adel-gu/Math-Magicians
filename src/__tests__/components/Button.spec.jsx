import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { Button } from '../../components/Button';

const mockGetBtnData = jest.fn();

beforeEach(() => {
  render(<Button data="0" getBtnData={mockGetBtnData} />);
});

describe('Button', () => {
  it('should render the button', () => {
    const buttonElement = screen.getByText(/0/i);

    expect(buttonElement).toHaveTextContent('0');
    expect(buttonElement).toBeInTheDocument();
  });

  it('should call mockGetBtnData with data', () => {
    const buttonElement = screen.getByText(/0/i);

    fireEvent.click(buttonElement);
    expect(mockGetBtnData).toBeCalledWith('0');
  });

  it('should match the snapshot', () => {
    const buttonElement = renderer.create(
      <Button data="0" getBtnData={mockGetBtnData} />,
    );

    expect(buttonElement).toMatchSnapshot();
  });
});
