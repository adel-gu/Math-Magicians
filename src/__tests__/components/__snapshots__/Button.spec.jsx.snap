// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match the snapshot 1`] = `
<button
  className="btn undefined"
  data-testid="buttonTestId"
  onClick={[Function]}
  style={
    Object {
      "gridColumn": "undefined / span 1",
      "gridRow": undefined,
    }
  }
>
  0
</button>
`;
